clippy_lints/src/matches.rs

```diff
@@ -322,6 +347,45 @@ fn check_overlapping_arms(cx: &LateContext, ex: &Expr, arms: &[Arm]) {
     }
 }
 
+fn check_wild_err_arm(cx: &LateContext, ex: &Expr, arms: &[Arm]) {
+    let ex_ty = walk_ptrs_ty(cx.tables.expr_ty(ex));
+    if match_type(cx, ex_ty, &paths::RESULT) {
+        for arm in arms {
+            if let PatKind::TupleStruct(ref path, ref inner, _) = arm.pats[0].node {
+                let path_str = print::to_string(print::NO_ANN, |s| s.print_qpath(path, false));
+                if_let_chain! {[
+                    path_str == "Err",
+                    inner.iter().any(|pat| pat.node == PatKind::Wild),
+                    let ExprBlock(ref block) = arm.body.node,
+                    is_panic_block(cx, block)
+                ], {
+                    // `Err(_)` arm with `panic!` found
+                    span_note_and_lint(cx,
+                                       MATCH_WILD_ERR_ARM,
+                                       arm.pats[0].span,
+                                       "Err(_) will match all errors, maybe not a good idea",
+                                       arm.pats[0].span,
+                                       "to remove this warning, match each error seperately \
+                                        or use unreachable macro");
+                }}
+            }
+        }
+    }
+}
+
+// If the block contains only a `panic!` macro (as expression or statement)
+fn is_panic_block(cx: &LateContext, block: &Block) -> bool {
+    match (&block.expr, block.stmts.len(), block.stmts.first()) {
+        (&Some(ref exp), 0, _) => {
+            is_expn_of(cx, exp.span, "panic").is_some() && is_expn_of(cx, exp.span, "unreachable").is_none()
+        },
+        (&None, 1, Some(ref stmt)) => {
```

Comment:
It's not compiling without the first `ref` in `ref exp`, but the second in `ref stmt` can be removed.

```
error[E0507]: cannot move out of borrowed content
   --> clippy_lints/src/matches.rs:379:10
    |
379 |         (&Some(exp), 0, _) => {
    |          ^^^^^^---^
    |          |     |
    |          |     hint: to prevent move, use `ref exp` or `ref mut exp`
    |          cannot move out of borrowed content
```